### PR TITLE
Add CMake support for C++ backend in test/generator

### PR DESCRIPTION
**Pull request by steven-johnson**
When the CMake rules were rewritten a while back, the support for building/testing generators with the C++ backend (instead of the standard LLVM, etc) got lost. This adds it back in.

Also made some drive-by fixes to the Makefile to enable some tests there that work correctly now.

Also made a drive-by fix in in Codegen_C to fix allocation nodes that were just wrappers around buffer_get_host -- this prevented the cleanup_on_error test from building with the C++ backend.

### CODE REVIEW

**Review by alexreinking — CHANGES_REQUESTED** [review]
One probable bug but otherwise LGTM

File: test/generator/CMakeLists.txt
Comment (steven-johnson) [reply test/generator/CMakeLists.txt:153]: oops. fixed.

**Comment by steven-johnson** [comment]
Adding @vksnk to review the changes in Codegen_C (especially the alignment change)

**Review by vksnk — APPROVED** [review]

**Comment by steven-johnson** [comment]
My previous changes were broken for wasm, so I ended up basically rewriting all the rules to be a bit clearer while fixing that, PTAL

**Comment by alexreinking** [comment]
> PTAL

@steven-johnson - I'm just going to wait until everything is green to review.

**Comment by steven-johnson** [comment]
>  I'm just going to wait until everything is green to review.

that's wise, given how annoying this has been to get right :-)

**Comment by steven-johnson** [comment]
Gentle ping to @alexreinking

File: test/generator/CMakeLists.txt
Comment (alexreinking) [inline test/generator/CMakeLists.txt:9]: Given that these values don't change throughout the file, it might be simpler to just write an if statement and set this value to a constant `1` or `0`

**Review by alexreinking — APPROVED** [review]
Good to get all this testing recovered! LGTM with one remark.

File: test/generator/CMakeLists.txt
Comment (steven-johnson) [reply test/generator/CMakeLists.txt]: Done

**Pull request merged**
